instance/models/mixins/redis.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2021 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app model mixins - Redis
+"""
+
+# Imports #####################################################################
+
+from functools import lru_cache
+import redis
+import json
+import string
+import urllib.parse
+
+from django.db import IntegrityError, models
+from django.utils.crypto import get_random_string
+import requests
+
+from instance.models.redis import RedisUser
+from instance.models.redis_server import RedisServer
+
+
+# Functions ###################################################################
+
+def new_redis_user(max_tries=10):
+    """
+    Return the primary key of a new Redis user.
+    """
+
+    if max_tries == 0:
+        raise IntegrityError("Cannot create unique redis user")
+
+    username = get_random_string(length=32, allowed_chars=string.ascii_lowercase)
+
+    if RedisUser.objects.filter(username=username).count() > 0:
+        return new_redis_user(max_tries - 1)
+
+    user = RedisUser.objects.create(
+        username=username,
+        password=get_random_string(length=64)
+    )
+
+    return user.pk
+
+
+def select_random_redis_server():
+    """
+    Helper for the field default of `redis_server`.
```

Comment:
Could you expand the docstring here to explain why this is required?

Reply:
I don't think it is required, wanted to write a todo comment here, but it seems I forgot. Actually that's how the rabbitmq server selection worked though I think that was a bad choice back in time to do this. Let's not follow that pattern. Will remove it.

Reply:
I checked the logic behind it. The select random selects a random server that accepts new connections and if the default instance redis URL is set, create a redis server based on it. Actually, we don't need the "random selection" logic, but the create default server logic is handy.

Reply:
What are your thoughts on this @swalladge ?

Reply:
Fine by me, as long as it's documented. :)

Reply:
Added more documentation as part of `cc739563`